### PR TITLE
OKTA-527800 Native Authentication POC.

### PR DESCRIPTION
**Pull request by JayNewstrom**
What this is:
- My current expectations on how things will be architected and emitted
- A work in progress that will be continually refactored moving forward

What this is not:
- Complete (in any way)
- To be released with the rest of the SDK
- Fully tested
- A UI SDK

### CODE REVIEW

File: native-authentication/src/main/java/com/okta/nativeauthentication/IdxResponseTransformer.kt [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+                        builder.elements.addAll(remediation.asAction())
+                    }
+                    producerScope.send(builder.build())
+                }
+            }
+        }
+    }
+
+    private fun IdxRemediation.Form.Field.asDynamicAuthFields(): List<Element.Builder<*>> {
+        return when (true) {
```
Comment (rajdeepnanua-okta) [inline native-authentication/src/main/java/com/okta/nativeauthentication/IdxResponseTransformer.kt:64]: Isn't this the same as `return when {`?

File: native-authentication/src/main/java/com/okta/nativeauthentication/form/Element.kt [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+            var label: String = ""
+            var value: String = ""
+            var isSecret: Boolean = false
+
+            override fun build(): TextInput {
+                return TextInput(idxField, label, isSecret, value)
+            }
+        }
+
+        @Volatile var value: String = value
```
Comment (rajdeepnanua-okta) [inline native-authentication/src/main/java/com/okta/nativeauthentication/form/Element.kt:57]: Why do we need to add volatile in this case?

File: native-authentication/src/main/java/com/okta/nativeauthentication/NativeAuthenticationClient.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        fun create(
+            callback: Callback,
+            interactionCodeFlowFactory: suspend () -> OidcClientResult<InteractionCodeFlow>,
+        ): Flow<Form> {
+            suspend fun ProducerScope<Form>.start() {
+                when (val result = interactionCodeFlowFactory()) {
+                    is OidcClientResult.Error -> {
+                        RetryFormBuilder.emit(this) {
+                            start()
```
Comment (rajdeepnanua-okta) [inline native-authentication/src/main/java/com/okta/nativeauthentication/NativeAuthenticationClient.kt:43]: Passing in producerScope to other functions might make it a bit harder to track emissions once more emitting functions are added. For future iterations, could we look into having only the channelFlow-owning class do the emissions, and the other classes emit events for the base class to consume? (to make it more of a unidirectional data flow)

**Review by rajdeepnanua-okta — COMMENTED** [review]
Looks good generally. Left a comment about my thoughts on passing producerScope to other classes. It's not a problem now, but I think it might become difficult to manage in the future if more classes start emitting directly to producerScope

File: native-authentication/src/main/java/com/okta/nativeauthentication/IdxResponseTransformer.kt (hunk 1 above)
Comment (JayNewstrom) [reply native-authentication/src/main/java/com/okta/nativeauthentication/IdxResponseTransformer.kt:64]: Looks like it! I didn't realize it was optional! Thanks!

File: native-authentication/src/main/java/com/okta/nativeauthentication/form/Element.kt (hunk 2 above)
Comment (JayNewstrom) [reply native-authentication/src/main/java/com/okta/nativeauthentication/form/Element.kt:57]: Element is capable of being written to and read from multiple threads. Basically I want to ensure the correct value is used. 

https://stackoverflow.com/a/106787 has some more information.

File: native-authentication/src/main/java/com/okta/nativeauthentication/NativeAuthenticationClient.kt (hunk 3 above)
Comment (JayNewstrom) [reply native-authentication/src/main/java/com/okta/nativeauthentication/NativeAuthenticationClient.kt:43]: Yeah. I agree with this. I'll plan to keep it as is for now. But I'll keep this comment in mind as I'm refactoring when adding in future features.

File: native-authentication-sample/src/main/java/com/okta/sample/auth/MainActivity.kt [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    val formState = forms.collectAsState(initial = null)
+    formState.value?.let {
+        FormUi(it)
+    }
+}
+
+@Composable
+fun FormUi(form: Form) {
+    val scrollState = rememberScrollState()
+    Column(
```
[the remark this answers is not in the document]
Comment (JayNewstrom) [reply native-authentication-sample/src/main/java/com/okta/sample/auth/MainActivity.kt:77]: In the final UI SDK, we _might_. For this simple sample, I'd say it's not worth the effort for right now.

**Review by rajdeepnanua-okta — APPROVED** [review]

**Comment by mikenachbaur-okta** [comment]
@jaynewstrom-okta I suggest you change this to target a feature branch, instead of `master`.

**Comment by JayNewstrom** [comment]
@mikenachbaur-okta Could you provide your reasoning? I don't see any reason to. It won't block any releases (it's not part of docs/promoting to maven central).

**Comment by mikenachbaur-okta** [comment]
I was thinking of leaving it in a separate branch until the design solidified more, particularly since it might undergo quite a bit of churn. To me, commits to the main branch should be more feature-complete, and this seems like it would undergo quite a bit of churn.

**Comment by JayNewstrom** [comment]
These are the types of commits that exist in both idx-android as well as mobile-kotlin, including the v2 (which was done in a feature branch, but merged to master). I think these commits are helpful when looking at the lines of code to see why things were done. When I did v2 for idx-kotlin in a feature branch, it caused more work and merge conflicts, which is why it's my vote to keep this against master.

**Pull request merged**